master/buildbot/util/git.py

```diff
@@ -270,53 +345,344 @@ def checkFeatureSupport(self):
 
     @defer.inlineCallbacks
     def _downloadSshPrivateKeyIfNeeded(self):
-        if self.sshPrivateKey is None:
-            return RC_SUCCESS
+        ssh_auth_warndeprecate(
+            "self._downloadSshPrivateKeyIfNeeded",
+            "self._git_auth.download_auth_files_if_needed",
+        )
+        workdir = self._get_auth_data_workdir()
+        value = yield self._git_auth.download_auth_files_if_needed(workdir)
+        return value
 
-        p = Properties()
-        p.master = self.master
-        private_key = yield p.render(self.sshPrivateKey)
-        host_key = yield p.render(self.sshHostKey)
-        known_hosts_contents = yield p.render(self.sshKnownHosts)
+    @defer.inlineCallbacks
+    def _removeSshPrivateKeyIfNeeded(self):
+        ssh_auth_warndeprecate(
+            "self._removeSshPrivateKeyIfNeeded",
+            "self._git_auth.remove_auth_files_if_needed",
+        )
+        workdir = self._get_auth_data_workdir()
+        value = yield self._git_auth.remove_auth_files_if_needed(workdir)
+        return value
+
+
+def ssh_auth_warndeprecate(old, new):
+    warn_deprecated(
+        "4.0.0",
+        f"{old} has been deprecated, "
+        f"please use {new}",
+    )
 
-        # not using self.workdir because it may be changed depending on step
-        # options
-        workdir = self._getSshDataWorkDir()
 
-        ssh_data_path = self._getSshDataPath()
-        yield self.runMkdir(ssh_data_path)
+class AbstractGitAuth(ComparableMixin):
+
+    compare_attrs = (
+        "ssh_private_key",
+        "ssh_host_key",
+        "ssh_known_hosts",
+    )
+
+    def __init__(
+        self,
+        ssh_private_key: IRenderable | None = None,
+        ssh_host_key: IRenderable | None = None,
+        ssh_known_hosts: IRenderable | None = None,
+    ) -> None:
+        self.did_download_auth_files = False
+
+        self.ssh_private_key = ssh_private_key
+        self.ssh_host_key = ssh_host_key
+        self.ssh_known_hosts = ssh_known_hosts
+
+        check_ssh_config('Git', self.ssh_private_key, self.ssh_host_key, self.ssh_known_hosts)
+
+    def is_auth_needed_for_git_command(self, git_command: str) -> bool:
+        if not git_command or self.ssh_private_key is None:
+            return False
+
+        git_commands_that_need_auth = [
+            'clone',
+            'fetch',
+            'ls-remote',
+            'push',
+            'submodule',
+        ]
+        if git_command in git_commands_that_need_auth:
+            return True
+        return False
+
+    @property
+    @abc.abstractmethod
```

Comment:
How are these abc.abstractmethod decorators useful? What hole do I have in my python knowledge? I always just override stuff in subclasses and that's it.

Reply:
Linter (pylint here) warn if a subclass does not override a method marked as abstract. I can remove it if preferred.

![image](https://github.com/buildbot/buildbot/assets/7926549/8ae21826-836c-4143-8733-fc59eec3baf6)

Reply:
Many pylint warnings are excessive. E.g. see https://github.com/buildbot/buildbot/blob/master/common/pylintrc for the list of pylint warnings that are disabled.

Reply:
In my experience I haven't yet seen a bug that is prevented by `@abc.abstractmethod`. I suggest we don't use it, because it's inconsistent across Buildbot codebase anyway.